### ISSUE TITLE
Improve accessibility

### ISSUE THREAD

**Issue by arantius** [greasemonkey/greasemonkey#2639]
I've received an email directly from a vision impaired user.  Accessibility of both the monkey menu and the script editor are currently poor and should be improved.

**Comment by Sxderp** [comment]
Like, screen reader accessibility? Could you provide more details?

**Comment by arantius** [comment]
Yes.  I don't have a lot more detail (yet), but I also don't want the task to be forgotten -- therefore this issue.

Best I can say for now:  Try closing your eyes and using GM 4.0.  How does it go?  Poorly I bet.  I'm guessing we need, at a minimum, some aria attributes all over the monkey menu.

**Comment by Sxderp** [comment]
Keyboard navigation may also be a good idea. At work there's someone I service that I know uses scripts / keyboard to do things because they don't have the dexterity to position the mouse. Currently I'm unable to tab into (after the menu is already opened)[1] nor navigate with arrow keys.

[1] As for opening the monkey menu, I don't think WebExtensions can modify that behavior at all. I could be missing an API though.

**Comment by arantius** [comment]
Yes, better keyboard navigation helps everybody.  Right now we're all <div> with onclick, nothing that tab would focus.  Perhaps we should be more <a> or similar?

I'm trying to/mostly like working "like a menu" so up/down arrows would make sense but I wonder how easy/hard that will be.

**Comment by arantius** [comment]
https://bugzilla.mozilla.org/show_bug.cgi?id=1324255

**Comment by Sxderp** [comment]
> Perhaps we should be more <a> or similar?
This would probably be better. I'm pretty sure the items would be highlight-able then. I recently created a Javascript gallery / pager in which you could switch between a gallery view or a photo paging view. I used <a> tags and anchors along with [onhashchange](https://developer.mozilla.org/en-US/docs/Web/API/WindowEventHandlers/onhashchange) in order to navigate between views and photo paging. Assuming that's still available in the popup it might be a viable option.

**Comment by arantius** [comment]
The above (merged) PR improves things, but needs work.

**Comment by jsl303** [comment]
Thank you for all the efforts to make Greasemonkey accessible.
For Windows 10 (starting 1703 Creators Update), you could use built-in screen reader called Narrator.
https://support.microsoft.com/en-us/help/22798/windows-10-narrator-get-started
For earlier Windows, you could test the accessibility using a free screen reader called nvda.
https://www.nvaccess.org/files/nvda/documentation/userGuide.html
For mac, you could test the accessibility using a built-in screen reader called VoiceOver.
https://help.apple.com/voiceover/info/guide/10.12/

**Comment by Sxderp** [comment]
There's a Bugzilla on focusing the popup window.
https://bugzilla.mozilla.org/show_bug.cgi?id=1324255

Looks promising and may make it for 59.

**Comment by Sxderp** [comment]
>  I assume screen reader users don't click buttons with the mouse?

Depends on the what they're using. There's a PhD student I work with (I work at a University, probably could have guessed by my email), who has difficulty with repetitive tasks (mouse movements and even typing). She uses various interaction methods like eye tracking (Tobii EyeX; interesting use of a device that's marketed to gamers) and voice commands (Dragon - something). Both of which can provide focus on windows and progressively zoom to highlight specific features. I don't know specifics but there's one piece of software that can find UI buttons with pretty good accuracy.

Anyway, the point is, as long as things are focusable, I'm pretty sure most capture systems are going to find them. [1] Once these software solutions find the elements they'll generally simulate mouse clicks for the user. Of course, keyboard navigation for individuals who aren't such extreme cases is welcome.

[1] Text boxes are really bad, don't use them. A lot of the software the student uses have a hard time with them (dunno why). Frames cause problems, but not as much.

**Comment by jsl303** [comment]
1. Open Firefox
2. Run Narrator by pressing control+windows+enter (in Windows 10).
3. Go to the search bar by pressing control+e
4. Move Narrator cursor by pressing capslock+right arrow until you get 
to Greasemoneky menu
5. Activate it by pressing capslock+enter
6. Press tab to move focus to the open menu
7. From there, you can just press up/down arrows to go through different 
options, and Narrator will read them.

**Comment by jcsteh** [comment]
See PRs #3058 and #3059.

**Comment by jcsteh** [comment]
> I still don't know how to test well. How do I open the monkey menu without clicking on it with the mouse? For that matter, how do I open the Firefox hamburger menu without the mouse?

In Firefox 67, [the browser toolbar buttons are now keyboard navigable](https://support.mozilla.org/en-US/kb/access-toolbar-functions-using-keyboard). In Firefox 70 (just released), we go a step further and you can now navigate to a button by typing the first few letters of its name. For example, to get to Greasemonkey:

1. Alt+d to focus address bar.
2. Tab to get to a toolbar button.
3. Type g r to jump focus to Greasemonkey.
4. Press space or enter.

> I assume screen reader users don't click buttons with the mouse?

Generally not. In web documents, screen readers provide a browse mode which allows users to read content which isn't keyboard navigable.
